File: src/day5.rs
use aoc_runner_derive::aoc;

use std::str::FromStr;

use anyhow::anyhow;
use lazy_static::lazy_static;
use regex::Regex;

#[derive(Debug)]
struct Step {
    count: usize,
    from: usize,
    to: usize,
}

impl FromStr for Step {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        lazy_static! {
            static ref RE: Regex = Regex::new(r#"move (\d+) from (\d+) to (\d+)"#).unwrap();
        }

        let caps = RE.captures(s).ok_or_else(|| anyhow!("bad move"))?;
        let count = caps[1].parse().unwrap();
        let from = caps[2].parse::<usize>().unwrap() - 1;
        let to = caps[3].parse::<usize>().unwrap() - 1;

        Ok(Self { count, from, to })
    }
}

#[derive(Debug)]
struct Input {
    stacks: Vec<Vec<u8>>,
    directions: Vec<Step>,
}

fn parse_stacks(input: &str) -> Vec<Vec<u8>> {
    let mut lines = input.lines().rev();

    let count = lines.next().unwrap().split_ascii_whitespace().count();
    let mut stacks = vec![vec![]; count];

    for l in lines.map(|l| l.as_bytes()) {
        for i in 0..count {
            let c = l[i * 4 + 1];
            if c != b' ' {
                stacks[i].push(c);
            }
        }
    }
    stacks
}

impl FromStr for Input {
    type Err = anyhow::Error;

    fn from_str(input: &str) -> Result<Self, Self::Err> {
        let halves: Vec<&str> = input.split("\n\n").collect();

        let stacks = parse_stacks(halves[0]);
        let directions = halves[1].lines().map(|l| l.parse().unwrap()).collect();

        Ok(Self { stacks, directions })
    }
}

fn as_char(ascii: u8) -> char {
    char::from_u32(ascii as u32).unwrap()
}

#[aoc(day5, part1)]
fn part1(i: &str) -> String {
    let mut input: Input = i.parse().unwrap();

    for step in input.directions {
        for _ in 0..step.count {
            let craned = input.stacks[step.from].pop().unwrap();
            input.stacks[step.to].push(craned);
        }
    }

    input
        .stacks
        .iter()
        .map(|s| as_char(*s.last().unwrap()))
        .collect()
}

#[aoc(day5, part2)]
fn part2(i: &str) -> String {
    let mut input: Input = i.parse().unwrap();

    for step in input.directions {
        let split_idx = input.stacks[step.from].len() - step.count;
        let craned = input.stacks[step.from].split_off(split_idx);
        input.stacks[step.to].extend(craned);
    }

    input
        .stacks
        .iter()
        .map(|s| as_char(*s.last().unwrap()))
        .collect()
}

File: hs/day5.hs
import Control.Lens((&), (%~), element)
import qualified Data.ByteString as BS(readFile)
import Data.List
import Data.Text(Text)
import qualified Data.Text as T
import Data.Text.Encoding(decodeUtf8)
import Data.Vector((!), Vector, fromList)
import Text.Regex.TDFA((=~))

data Step = Step {
    count :: Int,
    from :: Int,
    to :: Int
} deriving Show

data Input = Input {
    stacks :: Vector [Char],
    directions:: Vector Step
} deriving Show

parseInput :: Text -> Input
parseInput i = Input s d
    where
        halves = T.splitOn (T.pack "\n\n") i
        s = parseStacks . head $ halves
        d = parseDirections . last $ halves

parseStacks :: Text -> Vector [Char]
parseStacks st = fromList onStacks
    where
        -- Grab the last line.
        (lastLine:revLines) = reverse . T.lines $ st
        -- Split that last line; create an empty stack for each token ("word").
        emptyStacks = replicate (length . T.words $ lastLine) [] :: [[Char]]
        -- Iterate through every other line, appending it to the stacks.
        onStacks = foldl' addLine emptyStacks revLines
        -- For stack s at index i, append iff there's a crate there.
        addLine prev l = (\(s, i) -> pushCrate s $ crateChar l i) <$> zip prev [0..]
        pushCrate stack c = if c == ' ' then stack else c:stack
        -- Crates are formatted
        -- [a] [b] [c] [d]
        -- so char i*4 + 1 is the contents of the crate.
        crateChar :: Text -> Int -> Char
        crateChar l i = T.index l (i * 4 + 1)

parseDirections :: Text -> Vector Step
parseDirections d = fromList $ parseDirection <$> T.lines d
    where
        parseDirection = dirFromList . nums . T.unpack
        -- We'd better have three matches in the regex below.
        dirFromList [c, f, t] = Step c (f - 1) (t - 1)
        dirFromList _ = error "bad move"
        -- List of substring captures to list of ints
        nums l = read <$> substrs l :: [Int]
        -- Regex capture the numbers out of "move A from B to C".
        substrs l = s where (_, _, _, s) = reg l
        reg :: String -> (String, String, String, [String])
        reg l = l =~ "move ([0-9]+) from ([0-9]+) to ([0-9]+)"

part1 :: Input -> String
part1 i = foldl' (\acc s -> acc ++ [head s]) [] restacked
    where
        restacked = foldl' applyStep (stacks i) (directions i)
        -- Apply moveOneCrate `count stp` times
        applyStep :: Vector [Char] -> Step -> Vector [Char]
        applyStep onStacks stp = iterate (moveOneCrate stp) onStacks !! count stp
        -- Pop the top off of the `from stp`-th stack and push it to the `to stp`-th
        -- Lenses ftw? https://stackoverflow.com/a/15531874/713961
        moveOneCrate :: Step -> Vector [Char] -> Vector [Char]
        moveOneCrate stp stk = (stk & element (from stp) %~ tail) -- pop
            & element (to stp) %~ (head (stk ! from stp):) -- push


part2 :: Input -> String
part2 i = foldl' (\acc s -> acc ++ [head s]) [] restacked
    where
        restacked = foldl' applyStep (stacks i) (directions i)
        -- Pop n crates off the top of one stack and push them on the other.
        applyStep :: Vector [Char] -> Step -> Vector [Char]
        applyStep stk stp = (stk & element (from stp) %~ drop (count stp)) -- pop
            & element (to stp) %~ (movingCrates stk stp ++ ) -- push
        movingCrates :: Vector [Char] -> Step -> [Char]
        movingCrates stk stp = take (count stp) (stk ! from stp)

main :: IO ()
main = do
    input <- parseInput . decodeUtf8 <$> BS.readFile "../input/2022/day5.txt"
    print $ part1 input
    print $ part2 input
